### DISCUSSION TITLE
'data' vs. 'next'  - The spec and this library conflicts with apollographql/subscriptions-transport-ws and most clients

### DISCUSSION THREAD

**Discussion by AttilaTheFun** [enisdenjo/graphql-ws#288] [Q&A] [answered]
I'm doing some work routing the GraphQL subscriptions for my app through AWS API Gateway and I needed to refer to the graphql-ws spec. Unfortunately it seems this isn't well standardized and the spec in this repo was the closest thing I could find to an official spec. To make matters worse, one of the core operations differs between this library and the Apollo Client + Server libraries.

Apollo Server:
https://github.com/apollographql/subscriptions-transport-ws/blob/c6fb351fe3d7c7a447438798fab43c268b33036e/src/message-types.ts

Apollo iOS Client:
https://github.com/apollographql/apollo-ios/blob/550f1b11daef2dca71688b967f79e806d22826ce/Sources/ApolloWebSocket/OperationMessage.swift

Both the Apollo server and their clients use the operation type 'data' instead of 'next' when the server pushes GraphQL results to the client, and they don't recognize the 'next' event at all.

I'm not actually using this library, but I wanted to flag this discrepancy for you. If someone was to use this library with any of Apollo's client libraries, it likely wouldn't work. It might be worth adding a note about this discrepancy to the spec, too.

**Comment by enisdenjo** [comment, marked as answer]
The [GraphQL over WebSocket protocol](https://github.com/enisdenjo/graphql-ws/blob/master/PROTOCOL.md) this library uses is different from [the protocol `subscriptions-transport-ws` uses](https://github.com/apollographql/subscriptions-transport-ws/blob/master/PROTOCOL.md). Is also stated in the [readme disclaimer](https://github.com/enisdenjo/graphql-ws#disclaimer).

See also https://github.com/enisdenjo/graphql-ws/issues/154#issuecomment-819368765.

Additionally, if you are interested in using this protocol in AWS environment, I'd recommend checking:
- [`subscriptionless`](https://github.com/andyrichardson/subscriptionless)
- [`graphql-lambda-subscriptions`](https://github.com/reconbot/graphql-lambda-subscriptions)

**Reply by AttilaTheFun** [reply]
Ah interesting - I missed that before. Thanks for the explanation!
